### PR TITLE
Optimizing conj and assoc and making their behavior match clj's more closely

### PR DESCRIPTION
**Pull request by corasaurus-hex**
These appear to be significantly faster than the original functions and handle more corner cases in order to more closely match clojure's conj/assoc functions.

**EDIT**: I added get + assoc-in + assoc-in! as a bonus on this PR, and sped up str / added tests for it.

### CODE REVIEW

File: test/clava/compiler_test.cljs [hunk 1]
```diff
@@ -345,7 +345,33 @@
...
+    (is (eq [1 2 3 4] (jsv! '(conj [1 2 3] 4))))
+    (is (eq [1 2 3 4] (jsv! '(conj [1 2] 3 4)))))
+  (testing "sets"
+    (is (eq (js/Set. #js [1 2 3 4]) (jsv! '(conj #{1 2 3 4}))))
+    (is (eq (js/Set. #js [1 2 3 4]) (jsv! '(conj #{1 2 3} 4))))
+    (is (eq (js/Set. #js [1 2 3 4]) (jsv! '(conj #{1 2} 3 4)))))
+  #_(testing "maps"
+    (is (eq (js/Map. #js [["a" "b"] ["c" "d"]]) (jsv! '(conj #js {"a" "b" "c" "d"}))))
+    (is (eq (js/Map. #js [[1 2] [3 4]]) (jsv! '(conj {1 2} [3 4]))))
+    (is (eq (js/Map. #js [[1 2] [3 4] [5 6]]) (jsv! '(conj {1 2} [3 4] [5 6])))))
```
Comment (lilactown) [inline test/clava/compiler_test.cljs:363]: clavascript maps are turned into JS objects, not JS maps, which is one reason these tests don't work.

the other is that currently the compiler expects keys to always be keywords, and calls `namespace` on them without checking that they're something else (e.g. string). I'm not sure what we should do if a number is used, since technically object property names must always be strings.

Comment (corasaurus-hex) [reply test/clava/compiler_test.cljs]: in js you can use numbers for keys but they're automatically transformed into strings.

```
> a = {}
{}
> a[1] = 2
2
> a
{ '1': 2 }
> {[1]: 2}
{ '1': 2 }
```

**Comment by borkdude** [comment]
Appreciate this PR! Thanks a lot. I do prefer to work in small PRs that don't add a ton of stuff in one go, so it's easier to review.

About `assoc-in` we have a separate issue and I think we should first discuss if `assoc-in` should make clones of each object, or update them in place. This is an important decision that needs some thought instead of slipping it through in a PR I think?

About prettier: why is `let res = f(ctr,x);` better than `let res = f(ctr, x);`?
And `return n+1;` is changed into `return n + 1`, where is the consistency in that?
